src/session/ClientSession.h

```diff
@@ -0,0 +1,149 @@
+    void setSpace(SpaceInfo space) {
+        {
+            folly::RWSpinLock::WriteHolder wHolder(rwSpinLock_);
+            space_ = std::move(space);
+            session_.set_space_name(space_.name);
+        }
+    }
+
+    const std::string spaceName() {
+        folly::RWSpinLock::ReadHolder rHolder(rwSpinLock_);
+        return session_.get_space_name();
+    }
+
+    const std::string user() {
+        folly::RWSpinLock::ReadHolder rHolder(rwSpinLock_);
+        return session_.get_user_name();
+    }
+
+    std::unordered_map<GraphSpaceID, meta::cpp2::RoleType> roles() {
+        folly::RWSpinLock::ReadHolder rHolder(rwSpinLock_);
+        return roles_;
+    }
+
+    StatusOr<meta::cpp2::RoleType> roleWithSpace(GraphSpaceID space) {
+        folly::RWSpinLock::ReadHolder rHolder(rwSpinLock_);
+        auto ret = roles_.find(space);
+        if (ret == roles_.end()) {
+            return Status::Error("No role in space %d", space);
+        }
+        return ret->second;
+    }
+
+    bool isGod() {
+        folly::RWSpinLock::ReadHolder rHolder(rwSpinLock_);
+        // Cloud may have multiple God accounts
+        for (auto &role : roles_) {
+            if (role.second == meta::cpp2::RoleType::GOD) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    void setRole(GraphSpaceID space, meta::cpp2::RoleType role) {
+        folly::RWSpinLock::WriteHolder wHolder(rwSpinLock_);
+        roles_.emplace(space, role);
+    }
+
+    uint64_t idleSeconds();
+
+    void charge();
+
+    int32_t getTimezone() {
+        folly::RWSpinLock::ReadHolder rHolder(rwSpinLock_);
+        return session_.get_timezone();
+    }
+
+    void setTimezone(int32_t timezone) {
+        {
+            folly::RWSpinLock::WriteHolder wHolder(rwSpinLock_);
+            session_.set_timezone(timezone);
+            // TODO: if support ngql to set client's timezone,
+            //  need to update the timezone config to metad when timezone executor
+        }
+    }
+
+    void updateGraphAddr(const HostAddr &hostAddr) {
+        {
+            folly::RWSpinLock::WriteHolder wHolder(rwSpinLock_);
+            if (session_.get_graph_addr() == hostAddr) {
+                return;
+            }
+            session_.set_graph_addr(hostAddr);
+        }
+    }
+
+    const meta::cpp2::Session getSession() {
+        folly::RWSpinLock::ReadHolder rHolder(rwSpinLock_);
+        return session_;
+    }
+
+    void updateSpaceName(const std::string &spaceName) {
+        folly::RWSpinLock::WriteHolder wHolder(rwSpinLock_);
+        session_.set_space_name(spaceName);
+    }
+
+private:
+    ClientSession() = default;
+
+    explicit ClientSession(meta::cpp2::Session &&session, meta::MetaClient* metaClient);
+
+
+private:
+    SpaceInfo               space_;
+    time::Duration          idleDuration_;
+    meta::cpp2::Session     session_;
+    meta::MetaClient*       metaClient_{ nullptr};
+    folly::RWSpinLock       rwSpinLock_;
```

Comment:
There's some execution error if client runs different queries asynchronously. And it's better to mange client session in SesssionManager.

We should make sure that server doesn't run different queries of same session in parallel.

Reply:
Yes, this problem has always existed, and this problem will be solved by another PR.